fetch: Fix various issues related to `zig fetch --save`

castholm opened the pull request:
Closes #18639
Closes #21690

- `zig fetch --save file/path` is now an error.
- `--save`ing over an existing `path` entry will now correctly replace it with an `url` entry.
- Fetching by file URL (as in `zig fetch file:///C:/foo/bar.zip`) did not previously work, but does now.

The purpose of `zig fetch` is to copy a package to the global cache to avoid needing to access the network. It is not to manage your `build.zig.zon`. `--save` is provided as a convenience since calculating the hash by hand would be unreasonable, but it does not make sense to `--save` a package fetched by a (CWD-relative) file path since `path` entries in a build.zig.zon are for relative paths inside a package. `path` entries can only be added by manually editing your `build.zig.zon`.

Both `zig fetch` and `build.zig.zon` support fetching by file URL scheme, for example if a package is on an intranet network drive, but this did not previously always work due to a field being uninitialized and due to file URL-encoded absolute paths on Windows containing drive letters requiring extra attention. This PR fixes those bugs and improves validation to ensure file URLs are spec-compliant (based on RFC 8089 and https://url.spec.whatwg.org/).

I've verified the correctness of these changes by running the following commands locally (on Windows):
- `zig fetch --save ..\..\foo\` -> error (can't save when fetching by file path)
- `zig fetch --save ..\..\foo.zip` -> error (can't save when fetching by file path)
- `zig fetch --save file:///foo.zip` -> ok
- `zig fetch --save file:/foo.zip` -> ok
- `zig fetch --save file:///C:/foo.zip` -> ok
- `zig fetch --save file:/C:/foo.zip` -> ok
- `zig fetch --save file://localhost/C:/foo.zip` -> ok
- `zig fetch --save file:foo.zip` -> error (invalid file URL)

Additionally, it turns out that there were a lot of test declarations in `/src` that were unreferenced and untested, some of which were for the package manager internals. I de-bitrotted those tests and added them under a new `test-cli-internals` test set, which tests everything referenced by `/src/main.zig`.

castholm commented on src/Package/Fetch.zig:
```diff
@@ -65,6 +65,9 @@ oom_flag: bool,
 /// If the resource pointed to by the location is not a Git-repository, this
 /// will be left unchanged.
 latest_commit: ?git.Oid,
+/// If `location` was `path_or_url`, this will be set to indicate
+/// whether the CLI input was identified as a file path or a URL.
+actual_path_or_url_kind: enum { path, url },
```
I just want to note that this does mean that if the user invokes a command like `zig fetch --save ./foo.tar.gz`, the package will first be copied to the global cache before printing the error. This is obviously not perfect; ideally, the command should detect that the user passed a file path in combination with `--save` and exit before doing any work. However, if we want a maximally consistent and correct behavior, each solution comes with its own downsides

- Currently, we test if the `path_or_url` input is a directory we can open, then a file we can open, then finally a URL. Moving this logic to the `cmdFetch()` function before we actually fetch the package would mean we now have open file handles that we need to pass to the `run()` function and correctly close them if any errors occur on the way there (but not if `run()` returns an error since `run()` claims ownership of the handle). This is not trivial and will result in ugly code.
- We could change the order of operations and instead try parsing the input as an URL first. However, an input like `c:foo` (on Windows) or `http:foo` (on POSIX) would be interpreted as a URL despite possibly being a valid file path, so this isn't ideal. Even if we try to rule out values we know are not valid URLs that can be fetched, an input like `http://example.com/example.tar.gz` *could* be a valid file path to the subpath `example.com/example.tar.gz` inside the directory `http:`.
- We could also change the fetch command so that the user explicitly has to specify whether the input is a path or url, as in `zig fetch --path http:foo --save`.

I've thought about this a lot and I'm sure what the best way to handle this is, or if it's even worth putting energy into. I just think edge cases matter, however obscure they might be.

castholm replied:
4 months later and I realize I was seriously overthinking this. Just try to parse the path as a URL in advance (but detect Windows drive letters) and treat it as a path if it's not a valid URL. This is much better. For edge cases like a directory named `http:foo` the user can just pass `./http:foo` instead.

castholm commented on build.zig:
```diff
@@ -379,32 +379,6 @@ pub fn build(b: *std.Build) !void {
     const test_target_filters = b.option([]const []const u8, "test-target-filter", "Skip tests whose target triple do not match any filter") orelse &[0][]const u8{};
     const test_slow_targets = b.option(bool, "test-slow-targets", "Enable running module tests for targets that have a slow compiler backend") orelse false;
 
-    const test_cases_options = b.addOptions();
```
These were unreferenced.

castholm commented on build.zig:
```diff
@@ -531,6 +505,21 @@ pub fn build(b: *std.Build) !void {
+        .test_target_filters = test_target_filters,
+        .test_slow_targets = test_slow_targets,
+        .root_src = "src/main.zig",
+        .name = "cli-internals",
```
That would work too, so long as it's not misleading since it tests package manager internals and other things not strictly related to compiling code. I had trouble coming up with a suitable name in the first place, the main point I wanted to get across was that it tests `src/main.zig`, similar to how `test-std` tests `lib/std/std.zig`.

alexrp replied:
I'm mostly suggesting `test-compiler-internals` because we already have `test-cli`:

https://github.com/ziglang/zig/blob/e5d9d3f8a128ccf38b7e2f82c6abe70fb34e9eec/test/tests.zig#L1031-L1032

This new step runs all the unit tests defined within the compiler source code, right? In that case, I think `test-cli-internals` would be misleading given the existence of `test-cli` which narrowly tests the CLI.

castholm replied:
I changed it to `test-compiler-internals`.

castholm commented:
@andrewrk FYI, I rebased the branch and squashed the last two commits since they were already cherry-picked into #22222 and their previous messages no longer made sense. Your force push ran into a spurious CI failure so the checks needed to be rerun anyway.

andrewrk commented on src/Package/Fetch.zig:
```diff
@@ -274,14 +274,19 @@ pub const Location = union(enum) {
+    /// Indicates that the archive the URL refers to should
+    /// be fetched and copied into the global package cache.
+    /// Recursive Fetch tasks will never use this Location.
+    cli_url: std.Uri,
```
src/Package/Fetch.zig shouldn't know about the CLI.

castholm replied:
Please clarify, is your objection just to the names having "cli" in them, or that there are fields that are only used to serve the needs of `zig fetch` and thus only used by the CLi in practice? If its the names they can be changed, no problem, but if its the latter I don't really see a good way around it.

castholm replied:
Renamed them to just `url` and `path`, do you think that's better? Any other possible problems were present with the old `url_or_path` as well, all I did was break them out into two distinct fields.
